services/src/main/java/org/keycloak/storage/UserStorageManager.java

```diff
@@ -170,7 +182,54 @@ protected void deleteInvalidUser(final RealmModel realm, final UserModel user) {
             providersStream = Stream.concat(providersStream, Stream.of(federatedStorageProvider));
         }
 
-        return paginatedStream(providersStream.flatMap(pagedQuery::query), firstResult, maxResults);
+        final AtomicInteger currentFirst;
+
+        if (firstResult == null || firstResult <= 0) { // We don't want to skip any users so we don't need to do firstResult filtering
+            currentFirst = new AtomicInteger(0);
+        } else {
+            AtomicBoolean droppingProviders = new AtomicBoolean(true);
+            currentFirst = new AtomicInteger(firstResult);
+
+            providersStream = providersStream
+                .filter(provider -> { // This is basically dropWhile
+                    if (!droppingProviders.get()) return true; // We have already gathered enough users to pass firstResult number in previous providers, we can take all following providers
+
+                    long expectedNumberOfUsersForProvider = countQuery.query(provider, () -> 0, () -> currentFirst.get() + 1); // check how many users we can obtain from this provider
+
+                    if (expectedNumberOfUsersForProvider == currentFirst.get()) { // This provider provides exactly the amount of users we need for passing firstResult, we can set currentFirst to 0 and drop this provider
+                        currentFirst.set(0);
+                        droppingProviders.set(false);
+                        return false;
+                    }
+
+                    if (expectedNumberOfUsersForProvider > currentFirst.get()) { // If we can obtain enough enough users from this provider to fulfill our need we can stop dropping providers
+                        droppingProviders.set(false);
+                        return true; // don't filter out this provider because we are going to return some users from it
+                    }
+
+                    // This provider cannot provide enough users to pass firstResult so we are going to filter it out and change firstResult for next provider
+                    currentFirst.set((int) (currentFirst.get() - expectedNumberOfUsersForProvider));
+                    return false;
+                });
+        }
+
+        // Actual user querying
+        if (maxResults == null || maxResults < 0) {
+            // No maxResult set, we want all users
+            return providersStream
+                    .flatMap(provider -> pagedQuery.query(provider, () -> currentFirst.getAndSet(0), () -> null));
+        } else {
+            final AtomicInteger currentMax = new AtomicInteger(maxResults);
+
+            // Query users with currentMax variable counting how many users we return
+            return providersStream
+                    .filter(provider -> currentMax.get() != 0) // If we reach currentMax == 0, we can skip querying all following providers
+                    .flatMap(provider -> pagedQuery.query(provider, () -> currentFirst.getAndSet(0), currentMax::get))
```

Comment:
What would happen if the `firstSupplier` would be evaluated twice in the `pagedQuery.query` body?